service/src/main/java/bio/terra/workspace/service/workspace/flight/DeleteControlledSamResourcesStep.java

```diff
@@ -0,0 +1,82 @@
+package bio.terra.workspace.service.workspace.flight;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import bio.terra.workspace.service.resource.controlled.ControlledResource;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import java.util.List;
+
+/**
+ * A step which deletes the Sam data of the controlled resources provided by {@link
+ * ListControlledResourceIdsStep} Multiple deletions are run in parallel as resources are
+ * independent.
+ */
+public class DeleteControlledSamResourcesStep implements Step {
+
+  private final SamService samService;
+  private final AuthenticatedUserRequest userRequest;
+
+  public DeleteControlledSamResourcesStep(
+      SamService samService, AuthenticatedUserRequest userRequest) {
+    this.samService = samService;
+    this.userRequest = userRequest;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    FlightMap workingMap = flightContext.getWorkingMap();
+
+    @SuppressWarnings("unchecked")
+    List<ControlledResource> controlledResourceList =
+        workingMap.get(ControlledResourceKeys.CONTROLLED_RESOURCE_LIST, List.class);
+
+    try {
+      // Deleting each resource takes a call to Sam. As there may be many many resources and they
+      // are all independent, running these calls in parallel can save time.
+      controlledResourceList.parallelStream()
+          .forEach(
+              controlledResource -> {
```

Comment:
If your function returned `Optional<InterruptedException>` or an enum of succeeded/interrupted, then you could collect a map of controlled resource -> interruptedException and do something clever with that, like maybe an exception saying "3 of 99 controlled resources failed to delete".

Though, style-wise, I usually like to avoid side effects in `map` or other functional functions.

You could also try putting each delete call in a `CompletableFuture` and doing an `awaitAll()` on them. I haven't done that in a few years, and you still need to decide how to handle the exception. Why is it checked anyway, if it's almost always fatal?

Reply:
That's probably a question for Stairway that I don't know enough to answer - tabling this for now since I'm scrapping the parallelism